airbyte-cdk/python/unit_tests/sources/streams/http/requests_native_auth/test_requests_native_auth.py

```diff
@@ -0,0 +1,160 @@
+            client_id=TestOauth2Authenticator.client_id,
+            client_secret=TestOauth2Authenticator.client_secret,
+            refresh_token=TestOauth2Authenticator.refresh_token,
+        )
+
+        mocker.patch.object(Oauth2Authenticator, "refresh_access_token", return_value=("access_token", 1000))
+        header = oauth.get_auth_header()
+        assert {"Authorization": "Bearer access_token"} == header
+
+    def test_get_auth_header_expired(self, mocker):
+        """
+        Should retrieve new token if current token is expired.
+        """
+        oauth = Oauth2Authenticator(
+            token_refresh_endpoint=TestOauth2Authenticator.refresh_endpoint,
+            client_id=TestOauth2Authenticator.client_id,
+            client_secret=TestOauth2Authenticator.client_secret,
+            refresh_token=TestOauth2Authenticator.refresh_token,
+        )
+
+        expire_immediately = 0
+        mocker.patch.object(Oauth2Authenticator, "refresh_access_token", return_value=("access_token_1", expire_immediately))
+        oauth.get_auth_header()  # Set the first expired token.
+
+        valid_100_secs = 100
+        mocker.patch.object(Oauth2Authenticator, "refresh_access_token", return_value=("access_token_2", valid_100_secs))
+        header = oauth.get_auth_header()
+        assert {"Authorization": "Bearer access_token_2"} == header
+
+    def test_refresh_request_body(self):
+        """
+        Request body should match given configuration.
+        """
+        scopes = ["scope1", "scope2"]
+        oauth = Oauth2Authenticator(
+            token_refresh_endpoint=TestOauth2Authenticator.refresh_endpoint,
+            client_id=TestOauth2Authenticator.client_id,
+            client_secret=TestOauth2Authenticator.client_secret,
+            refresh_token=TestOauth2Authenticator.refresh_token,
+            scopes=scopes,
+        )
+        body = oauth.get_refresh_request_body()
+        expected = {
+            "grant_type": "refresh_token",
+            "client_id": "client_id",
+            "client_secret": "client_secret",
+            "refresh_token": "refresh_token",
+            "scopes": scopes,
+        }
+        assert body == expected
+
+    def test_refresh_access_token(self, mocker):
+        oauth = Oauth2Authenticator(
+            token_refresh_endpoint=TestOauth2Authenticator.refresh_endpoint,
+            client_id=TestOauth2Authenticator.client_id,
+            client_secret=TestOauth2Authenticator.client_secret,
+            refresh_token=TestOauth2Authenticator.refresh_token,
+        )
+        resp = Response()
+        resp.status_code = 200
+
+        mocker.patch.object(requests, "request", return_value=resp)
+        mocker.patch.object(resp, "json", return_value={"access_token": "access_token", "expires_in": 1000})
+        token = oauth.refresh_access_token()
+
+        assert ("access_token", 1000) == token
+
+    def test_auth_call_method(self, mocker):
```

Comment:
How are we testing the `__call__` method? shouldn't there be tests for each authenticator which take a `request` and call the authenticator on it `authenticator(request)` and compare the before-and-after of the request object?

Reply:
But the current implementation is indeed tests the `__call__ ` method.
Here is the wrapper, after which the `__call__` method is triggered https://github.com/psf/requests/blob/main/requests/models.py#L558.
I've added this implementation, because here we cover the whole request flow from the beginning to actual sending.
But the `send` method is mocked to get the data what we need to test and not actually send the request.